### PR TITLE
[SYCL] Add a test for alignment of allocated USM memory

### PR DESCRIPTION
**Pull request by aelovikov-intel**

### CODE REVIEW

File: SYCL/USM/alloc_functions.cpp [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        [&](auto... Fs) {
+          int Case = 0;
+          (void)std::initializer_list<int>{
+              (check(Expected, Fs, Line, Case++), 0)...};
+        },
+        Funcs);
+  };
+
+  auto MDevice = [&](auto... args) {
+    return malloc_device(sizeof(Aligned), args...);
```
Comment (gmlueck) [inline SYCL/USM/alloc_functions.cpp]: Kind of a nit, but this is only guaranteed to be aligned at `alignof(std::max_align_t)` if the allocation size is at least as big as `sizeof(std::max_align_t)`.  Note this statement in the spec:

> Pointer is suitably aligned for any object with fundamental alignment **whose size is less than or equal to the requested allocation size**

Maybe you could define `Aligned` like this:

```
struct alignas(Align) Aligned {
  char c[sizeof(std::max_align_t)];
};
```

Comment (aelovikov-intel) [reply SYCL/USM/alloc_functions.cpp]: `sizeof(Aligned)` is `Align`: https://godbolt.org/z/337r5brns

However, you're right that those non-template allocations have nothing to do with the `Aligned` type itself. I'm changing the `sizeof` parameter to decouple testing those cases from the `Aligned` type.

**Review by gmlueck — APPROVED** [review]

**Comment by aelovikov-intel** [comment]
@v-klochkov , ping.

File: SYCL/USM/alloc_functions.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  auto TShared = [&](auto... args) {
+    return malloc_shared<Aligned>(1, args...);
+  };
+  CheckAll(Align, std::tuple{[&]() { return TShared(q); },
+                             [&]() { return TShared(d, ctx); }});
+
+  auto ATDevice = [&](size_t Align, auto... args) {
+    return aligned_alloc_device<Aligned>(Align, 1, args...);
+  };
+  CheckAll(Align, std::tuple{[&]() { return ATDevice(Align / 2, q); },
```
Comment (v-klochkov) [inline SYCL/USM/alloc_functions.cpp:165]: At this place and few other as well: why do you request alignment = "Align/2" and expect the actual alignment would be twice bigger, i.e. "Align" ?

Comment (aelovikov-intel) [reply SYCL/USM/alloc_functions.cpp:165]: From table 94 in https://github.com/gmlueck/SYCL-Docs/blob/8c37a7eb9483938f35cace000c11be12ea77538e/adoc/chapters/programming_interface.adoc#usm-allocations:

<html>
<body>


Alignment parameter alignment specified Templated on allocation type T | Pointer is suitably aligned for an object of type T or it is aligned to the specified alignment, whichever is greater.
-- | --



</body>
</html>

The `Align / 2` is less than the template type parameter's alignment so the latter must be used.

**Comment by aelovikov-intel** [comment]
ping @v-klochkov

**Comment by v-klochkov** [comment]
Ex

> ping @v-klochkov

Sorry for the delay, I'll finish review before the end of today.

**Review by v-klochkov — APPROVED** [review]

File: SYCL/USM/alloc_functions.cpp (hunk 2 above)
Comment (v-klochkov) [reply SYCL/USM/alloc_functions.cpp:165]: I see now, thank you for the link.

**Pull request merged**